dilated-file-transfer.md

```diff
@@ -0,0 +1,499 @@
+# Dilated File-Transfer Protocol
+
+This version of the file-transfer protocol is a complete replacement for the original (referred to as "classic") file-transfer protocol.
+
+Both sides must support and use Dilation (see `dilation-protocol.md`).
+
+Any all-caps words ("MAY", "MUST", etc) follow RFC2119 conventions.
+
+    NOTE: there are several open questions / discussion points, some with corresponding "XXX" comments inline. See also [Discussion and Open Questions](#discussion)
+
+
+## Philosophic Note
+
+The place to be opinionated is in client implementations.
+The place to be flexible is in protocol specifications.
+
+Everyone reading this likely has a particular sort of user experience (UX) in mind; the protocol should _allow_ reasonable features but shouldn't _demand_ any particular UX.
+
+The protocol absolutely MUST be extensible in the future (we can't do everything correctly immediately).
+
+
+## Overview and Features
+
+This specification is an application-level Magic Wormhole protocol defining a flexible, "session"-based approach to file transfer.
+Dilation must be supported by both clients (see :ref:`dilation-protocol.md`).
+Client implementations can allow either side to offer files/directories to send while the other side may accept or reject each offer.
+Either side MAY terminate the transfer session (by closing the wormhole).
+Either side MAY select a one-way mode, similar to the classic protocol.
+
+Files or directories are offered and sent individually, with no dependency on zip or other archive formats.
+
+Metadata is included in the offers to allow the receiver to decide if they want that file before the transfer begins.
+
+"Offers" generally correspond to what a user might select; a single-file offer is possible but so is a directory.
+In both cases, they are treated as "an offer" although a directory may consist of dozens or more individual files.
+
+XXX include compression in this revision, or allow that to be a future enhancement?
+
+
+## Version Negotiation
+
+There is an existing file-transfer protocol which does not use Dilation (called "classic" in this document).
+Clients supporting newer versions of file-transfer (i.e. the one in this document) SHOULD offer backwards compatibility.
+
+In the mailbox protocol, applications can indicate version information.
+The classic file-transfer protocol doesn't use this feature so the version information is empty.
+This new protocol will include a dict like:
+
+```json
+{
+    "transfer-v1": {
+        "mode": "{send|receive|connect}",
+        "features": [],
+    }
+}
+```
+
+The version of the protocol is the `"-v1"` tag in `"transfer-v1"`.
+A peer supporting newer versions may include `"transfer-v2"` or `"transfer-v3"` etc.
+There is currently only one version: `1`.
```

Comment:
Why is this version 1? Wouldn't it make more sense to make this version 2 as there already is a version "1" even though not specified? I think this will create lots of confusion between "original file transfer protocol" and this "version 1" which is actually the second iteration of the file transfer protocol.

Reply:
Sure we could start at v2 if that makes more sense. I guess I looked at it as "the version of the Dilated File Transfer protocol".

(There would just never be a `transfer-v1` in the application version information then, but okay).